sw/cheri/common/platform-pinmux.hh

```diff
@@ -0,0 +1,360 @@
+/**
+ * Copyright lowRISC contributors.
+ * Licensed under the Apache License, Version 2.0, see LICENSE for details.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+#include <debug.hh>
+#include <stdint.h>
+#include <utils.hh>
+
+/**
+ * A driver for Sonata's Pin Multiplexer (Pinmux).
+ *
+ * This driver can be used to select which block output is placed on a given
+ * output pin, and to select which input pin is provided to a given block
+ * input. Sonata's pinmux only allows certain selections per output pin / block
+ * input, which this driver describes.
+ *
+ * Documentation source can be found at:
+ * https://github.com/lowRISC/sonata-system/blob/6914b30a2769f881a53afb912032cd65a6d14307/doc/ip/pinmux.md
+ *
+ * Rendered documentation is served from:
+ * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
+ */
+class SonataPinmux : private utils::NoCopyNoMove {
+  /**
+   * Flag to set when debugging the driver for UART log messages.
+   */
+  static constexpr bool DebugDriver = false;
+
+  /**
+   * Helper for conditional debug logs and assertions.
+   */
+  using Debug = ConditionalDebug<DebugDriver, "Pinmux">;
+
+  /**
+   * A pointer/capability to the Pin Multiplexer's registers, where
+   * each sequential byte potentially corresponds to some mapped pinmux
+   * selector used to select the pin configuration.
+   */
+  volatile uint8_t *registers;
+
+ public:
+  /**
+   * The Output Pins defined for the Sonata board. These output pins can be
+   * multiplexed, meaning that they can be changed to output the outputs
+   * of different blocks (or disabled). The block outputs that can be
+   * selected are limited, and vary on a per-pin basis.
+   *
+   * Documentation sources:
+   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
+   * https://github.com/lowRISC/sonata-system/blob/4b72d8c07c727846c6ccb27754352388f3b2ac9a/data/pins_sonata.xdc
+   * https://github.com/newaetech/sonata-pcb/blob/649b11c2fb758f798966605a07a8b6b68dd434e9/sonata-schematics-r09.pdf
+   */
+  enum class OutputPin : uint16_t {
```

Comment:
Don't need to do this now but we should look at auto-generating these from the pinmux configuration

Reply:
I was considering this whilst doing it. At the very least it would be nice to have some matching names in the auto-generated documentation so that users can more easily identify what is muxable where. This would have to be a separate script though I think, as this driver is intended to eventually live in upstream `cheriot-rtos` and not in this repo.